proposals/extension_get_contexts.md

```diff
@@ -0,0 +1,171 @@
+our browser design. Due to this we cannot provide access to the JS context for
+these views, but we can allow an extension to query for them and potentially
+target them for messaging purposes.
+
+## Solution
+
+Considering the above situation, we’d like to propose a new (Manifest V3 only)
+method to asynchronously provide metadata about the view that is still useful
+for an extension: `extension.getContexts()`. This will allow extension service
+workers (or other contexts) to identify the active contexts in the extension.
+For example, this can be used to target messages to send using
+<code>[runtime.sendMessage()](https://developer.chrome.com/docs/extensions/reference/runtime/#method-sendMessage)</code>,
+etc.). Introducing this API will allow an easier migration from MV2.
+
+## API Proposal
+
+This new API would use a similar parameter signature as
+`extension.getViews()`but instead accept a new `contextProperties` object.
+
+```
+extension.getContexts()(
+  contextProperties?: object,
+)
+```
+
+`contextProperties` will have similar behavior to `extension.getViews()`’s
+<code>[fetchProperties](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)</code>
+object in that it will limit the scope of the query to certain tabs, context
+types, and windows. It would be defined as:
+
+```
+extension.ContextProperties = {
+  tabId? = int,                // Find context by tab id, omitted returns all tabs.
+  contextType? = ContextType,  // Find context by type, omitted returns all types.
+  windowId? = int,             // Find context by window id, omitted returns all
+                               // windows.
+}
+```
+
+`ContextType` would serve a similar purpose of
+<code>[fetchProperties](https://developer.chrome.com/docs/extensions/reference/extension/#method-getViews)</code>’s
+[ViewType](https://developer.chrome.com/docs/extensions/reference/extension/#type-ViewType)
+in order to limit the type of extension contexts returned by the API. It would
+be defined as:
+
+```
+extension.ContextType = {
+  TAB: 'TAB',                               // Tabs the extension is running in.
+  POPUP: 'POPUP',                           // Popups created by the extension.
+  SERVICE_WORKER: 'SERVICE_WORKER',         // Service workers for the extension.
+  OFFSCREEN_DOCUMENT: 'OFFSCREEN_DOCUMENT'  // Offscreen documents for the extension.
+};
+```
+
+It would return an array of new `Context` objects for each frame the extension
+is running in:
+
+```
+extension.Context = {
+  tabId: int,                    // ID of the tab for this view.
+  windowId: int,                 // ID of the window this tab is located in.
+  documentId: int,               // ID of the DOM document loaded in the tab.
+  frameId: int,                  // ID of the frame this context is running for.
+  contextType: ContextType,      // Context type -- tab, popup, etc.
+  url: string,                   // The current URL of the tab.
```

Comment:
Is this the tab URL or the frame's URL? Will is return or non-tab contexts, like popup or background?

Reply:
Corrected the typo, this would be the frame's URL so I renamed it to `frameUrl`.  In non-tab context like popup (Chromium as an example) popup  would be "chrome-extension://<id>/popup.html".

For a background page or service worker (since there's only one per extension) it doesn't seem helpful to know URL for it (e.g. "chrome-extension://<id>/background.js" so in those cases this would be empty. However, we could include it if there's a need for it? Applying this same logic would mean the field would be empty for content scripts as well, if supported.